File: javascript/data-structures/p271_HashTableProbing.js
/*
    CLRS Section 11.4, p. 271 and Exercise 11.4-1, p. 277
*/

class HashTableProbing {
    constructor(m = 0, colRes = 'lp') {
        this.m = m;
        this.colRes = colRes;

        this.T = new Array(m).fill(null);
        this.collisionsIns = 0;
        this.collisionsSrch = 0;
        this.collisionsDel = 0;
    }

     // auxiliary hash functions (cf. Exercise 11.4-1, p. 277)
    hAux1 = (k) => k;
    hAux2 = (k) => 1 + (k % (this.m - 1));

    // hash functions (cf. Exercise 11.4-1, p. 277)
    hLP = (k, i) => (this.hAux1(k) + i) % this.m; // linear probing
    hQP = (k, i) => { // quadratic probing
        const c1 = 1, c2 = 3;
        return (this.hAux1(k) + c1 * i + c2 * i * i) % this.m;
    }
    hDH = (k, i) => (this.hAux1(k) + i * this.hAux2(k)) % this.m; // double hashing

    // hash function selector by collision resolution strategy
    hSelector = (colRes, k, i) => {
        switch (colRes) {
            case 'lp':
                return this.hLP(k, i);
            case 'qp':
                return this.hQP(k, i);
            case 'dh':
                return this.hDH(k, i);
            default:
                break;
        }
    }

    resetCols = () => {
        this.collisionsDel = 0;
        this.collisionsIns = 0;
        this.collisionsSrch = 0;
    }

    hashInsert = (k) => {
        let i = 0;
        do {
            let j = this.hSelector(this.colRes, k, i);
            if (this.T[j] === null || this.T[j] === 'DELETED') {
                this.T[j] = k;
                return j;
            } else {
                i++;
                this.collisionsIns++;
            }
        } while (i !== this.m)
        throw new Error('hash table overflow')
    }

    hashSearch = (k) => {
        let i = 0;
        do {
            let j = this.hSelector(this.colRes, k, i);
            if(this.T[j] === k) {
                return j;
            }
            i++;
            this.collisionsSrch++;
        } while (i !== this.m)
        throw new Error('element not found')
    }

    hashDelete = (k) => {
        let i = 0, j;
        do {
            j = this.hSelector(this.colRes, k, i);
            if (this.T[j] === k) {
                this.T[j] = 'DELETED'
                return j;
            } else {
                i++;
                this.collisionsDel++;
            }
        } while (this.T[j] !== null && i !== this.m)
        throw new Error('element not found')
    }
}

module.exports = {
    HashTableProbing
}

File: python/data_structures/p271_hash_table_probing.py
'''
    CLRS Section 11.4, p. 271 and Exercise 11.4-1, p. 277
'''

class HashTableProbing:
    def __init__(self, m=0, col_res='lp'):
        self.m = m
        self.col_res = col_res
        self.T = [None for element in range(0, m)]
        self.collisions_ins = 0
        self.collisions_srch = 0
        self.collisions_del = 0

    # auxiliary hash functions (cf. Exercise 11.4-1, p. 277)
    def h_aux1(self, k):
        return k
    
    def h_aux2(self, k):
        return 1 + (k % (self.m - 1))
    
    # hash functions (cf. Exercise 11.4-1, p. 277)
    def h_lp(self, k, i): # linear probing
        return (self.h_aux1(k) + i) % self.m
    
    def h_qp(self, k, i): # quadratic probing
        c1 = 1; c2 = 3
        return (self.h_aux1(k) + c1 * i + c2 * i * i) % self.m
    
    def h_dh(self, k, i): # double hashing
        return (self.h_aux1(k) + i * self.h_aux2(k)) % self.m
    
    # hash function selector by collision resolution strategy
    def h_selector(self, col_res, k, i):
        selector = {
            'lp': self.h_lp,
            'qp': self.h_qp,
            'dh': self.h_dh
        }
        return selector[col_res](k, i)
    
    def reset_cols(self):
        self.collisions_del = 0
        self.collisions_ins = 0
        self.collisions_srch = 0
    
    def hash_insert(self, k):
        i = 0
        while True:
            j = self.h_selector(self.col_res, k, i)
            if self.T[j] is None or self.T[j] == 'DELETED':
                self.T[j] = k
                return j
            else:
                i += 1
                self.collisions_ins += 1
            if i == self.m:
                raise Exception('hash table overflow')
    
    def hash_search(self, k):
        i = 0
        while True:
            j = self.h_selector(self.col_res, k, i)
            if self.T[j] == k:
                return j
            i += 1
            self.collisions_srch += 1
            if i == self.m:
                raise Exception('element not found')
    
    def hash_delete(self, k):
        i = 0
        while True:
            j = self.h_selector(self.col_res, k, i)
            if self.T[j] == k:
                self.T[j] = 'DELETED'
                return j
            else:
                i += 1
                self.collisions_del += 1
            if self.T[j] is None and i == self.m:
                raise Exception('element not found')
